pytorch_lightning/core/lightning.py

```diff
@@ -1166,16 +1166,44 @@ def toggle_optimizer(self, optimizer: Optimizer, optimizer_idx: int):
 
         Override for your own behavior
 
+        It works with `untoggle_optimizer` to make sure param_requires_grad_state is properly reset.
+
         Args:
             optimizer:
             optimizer_idx:
         """
-        for param in self.parameters():
-            param.requires_grad = False
+        param_requires_grad_state = {}
+        for opt_idx, opt in enumerate(self.optimizers(use_pl_optimizer=False)):
+            if optimizer_idx != opt_idx:
+                for group in opt.param_groups:
+                    for param in group['params']:
+                        param_requires_grad_state[param] = param.requires_grad
+                        param.requires_grad = False
 
         for group in optimizer.param_groups:
             for param in group['params']:
-                param.requires_grad = True
+                if param in param_requires_grad_state:
+                    param.requires_grad = param_requires_grad_state[param]
+
+        self._param_requires_grad_state = param_requires_grad_state
+
+    def untoggle_optimizer(self, optimizer_idx: int):
+        """
+        .. note:: Only called when using multiple optimizers
+
+        Override for your own behavior
+
+        Args:
+            optimizer_idx:
+        """
+        for opt_idx, opt in enumerate(self.optimizers(use_pl_optimizer=False)):
+            if optimizer_idx != opt_idx:
```

Comment:
shouldn't untoggle be done for all optimizers?

Reply:
No, as toggle won't touch the current one.